File: mobile/android/app/src/main/java/com/example/seniordesignmobileapp/recording/SessionRecorder.kt
package com.example.seniordesignmobileapp.recording

import android.content.Context
import com.example.seniordesignmobileapp.model.ActiveSensorStatus
import com.example.seniordesignmobileapp.model.ImuSample
import com.example.seniordesignmobileapp.model.NetworkStatus
import com.example.seniordesignmobileapp.model.SavedSessionSummary
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.coroutines.withContext
import org.json.JSONArray
import org.json.JSONObject
import java.io.BufferedWriter
import java.io.File
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale

data class RecordingSessionInfo(
    val fileName: String,
    val absolutePath: String,
    val startedAtEpochMs: Long,
)

class SessionRecorder(
    context: Context,
) {
    private val sessionsDirectory = File(context.filesDir, "sessions").apply { mkdirs() }
    private val mutex = Mutex()
    private var activeSession: ActiveSession? = null

    suspend fun startSession(deviceName: String): RecordingSessionInfo =
        withContext(Dispatchers.IO) {
            mutex.withLock {
                check(activeSession == null) { "A recording session is already active." }

                val startedAtEpochMs = System.currentTimeMillis()
                val timestamp = FILE_NAME_FORMAT.format(Date(startedAtEpochMs))
                val fileName = "session-$timestamp.jsonl"
                val file = File(sessionsDirectory, fileName)
                val writer = file.bufferedWriter()
                activeSession = ActiveSession(file = file, writer = writer, startedAtEpochMs = startedAtEpochMs)

                writer.appendLine(
                    JSONObject()
                        .put("type", "session_start")
                        .put("recorded_at_epoch_ms", startedAtEpochMs)
                        .put("device_name", deviceName)
                        .toString()
                )
                writer.flush()

                RecordingSessionInfo(
                    fileName = file.name,
                    absolutePath = file.absolutePath,
                    startedAtEpochMs = startedAtEpochMs,
                )
            }
        }

    suspend fun appendSample(
        recordedAtEpochMs: Long,
        receivedAtElapsedMs: Long,
        sample: ImuSample,
    ): Boolean = appendEvent(
        JSONObject()
            .put("type", "sample")
            .put("recorded_at_epoch_ms", recordedAtEpochMs)
            .put("received_at_elapsed_ms", receivedAtElapsedMs)
            .put("sample", sample.toJson()),
    )

    suspend fun appendNetworkStatus(
        recordedAtEpochMs: Long,
        receivedAtElapsedMs: Long,
        status: NetworkStatus,
    ): Boolean = appendEvent(
        JSONObject()
            .put("type", "network_status")
            .put("recorded_at_epoch_ms", recordedAtEpochMs)
            .put("received_at_elapsed_ms", receivedAtElapsedMs)
            .put("network_status", status.toJson()),
    )

    suspend fun stopSession(reason: String): RecordingSessionInfo? =
        withContext(Dispatchers.IO) {
            mutex.withLock {
                val session = activeSession ?: return@withLock null
                val stoppedAtEpochMs = System.currentTimeMillis()
                session.writer.appendLine(
                    JSONObject()
                        .put("type", "session_end")
                        .put("recorded_at_epoch_ms", stoppedAtEpochMs)
                        .put("reason", reason)
                        .toString()
                )
                session.writer.flush()
                session.writer.close()
                activeSession = null

                RecordingSessionInfo(
                    fileName = session.file.name,
                    absolutePath = session.file.absolutePath,
                    startedAtEpochMs = session.startedAtEpochMs,
                )
            }
        }

    suspend fun listSessions(limit: Int = 10): List<SavedSessionSummary> =
        withContext(Dispatchers.IO) {
            sessionsDirectory
                .listFiles()
                .orEmpty()
                .filter { file -> file.isFile && file.extension == "jsonl" }
                .sortedByDescending(File::lastModified)
                .take(limit)
                .map { file ->
                    SavedSessionSummary(
                        fileName = file.name,
                        absolutePath = file.absolutePath,
                        sizeBytes = file.length(),
                        modifiedAtEpochMs = file.lastModified(),
                    )
                }
        }

    fun getSessionFile(fileName: String): File? {
        val candidate = File(sessionsDirectory, fileName)
        return candidate.takeIf { it.exists() && it.isFile }
    }

    private suspend fun appendEvent(event: JSONObject): Boolean =
        withContext(Dispatchers.IO) {
            mutex.withLock {
                val session = activeSession ?: return@withLock false
                session.writer.appendLine(event.toString())
                session.writer.flush()
                true
            }
        }

    private data class ActiveSession(
        val file: File,
        val writer: BufferedWriter,
        val startedAtEpochMs: Long,
    )

    private companion object {
        val FILE_NAME_FORMAT = SimpleDateFormat("yyyyMMdd-HHmmss", Locale.US)
    }
}

private fun ImuSample.toJson(): JSONObject =
    JSONObject()
        .put("version", version)
        .put("sensor_id", sensorId)
        .put("seq", seq)
        .put("timestamp_ms", timestampMs)
        .put("ax", ax)
        .put("ay", ay)
        .put("az", az)
        .put("gx", gx)
        .put("gy", gy)
        .put("gz", gz)

private fun NetworkStatus.toJson(): JSONObject =
    JSONObject()
        .put("version", version)
        .put("uptime_ms", uptimeMs)
        .put("active_sensor_count", activeSensorCount)
        .put(
            "sensors",
            JSONArray().apply {
                sensors.forEach { put(it.toJson()) }
            }
        )

private fun ActiveSensorStatus.toJson(): JSONObject =
    JSONObject()
        .put("sensor_id", sensorId)
        .put("seq", seq)
        .put("age_ms", ageMs)

File: mobile/ios/FH07Mobile/Recording/SessionRecorder.swift
import Foundation

struct RecordingSessionInfo: Equatable {
    let fileName: String
    let absolutePath: String
    let startedAtEpochMs: Int64
}

actor SessionRecorder {
    private let sessionsDirectory: URL
    private var activeSession: ActiveSession?

    private static let fileNameFormatter: DateFormatter = {
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyyMMdd-HHmmss"
        formatter.locale = Locale(identifier: "en_US_POSIX")
        formatter.timeZone = TimeZone.current
        return formatter
    }()

    init() {
        let documents = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask).first!
        self.sessionsDirectory = documents.appendingPathComponent("sessions", isDirectory: true)
        try? FileManager.default.createDirectory(at: sessionsDirectory, withIntermediateDirectories: true)
    }

    func startSession(deviceName: String) throws -> RecordingSessionInfo {
        if activeSession != nil {
            throw RecorderError.alreadyActive
        }

        let startedAtEpochMs = Int64(Date().timeIntervalSince1970 * 1000)
        let timestamp = Self.fileNameFormatter.string(from: Date(timeIntervalSince1970: TimeInterval(startedAtEpochMs) / 1000))
        let fileName = "session-\(timestamp).jsonl"
        let fileURL = sessionsDirectory.appendingPathComponent(fileName)

        FileManager.default.createFile(atPath: fileURL.path, contents: nil)
        guard let handle = try? FileHandle(forWritingTo: fileURL) else {
            throw RecorderError.cannotOpenFile
        }
        let session = ActiveSession(file: fileURL, handle: handle, startedAtEpochMs: startedAtEpochMs)
        activeSession = session

        let startEvent: [String: Any] = [
            "type": "session_start",
            "recorded_at_epoch_ms": startedAtEpochMs,
            "device_name": deviceName
        ]
        try writeJsonLine(handle: handle, object: startEvent)

        return RecordingSessionInfo(
            fileName: fileURL.lastPathComponent,
            absolutePath: fileURL.path,
            startedAtEpochMs: startedAtEpochMs
        )
    }

    func appendSample(recordedAtEpochMs: Int64, receivedAtElapsedMs: Int64, sample: ImuSample) -> Bool {
        let event: [String: Any] = [
            "type": "sample",
            "recorded_at_epoch_ms": recordedAtEpochMs,
            "received_at_elapsed_ms": receivedAtElapsedMs,
            "sample": sample.toJsonDictionary()
        ]
        return appendEvent(event)
    }

    func appendNetworkStatus(recordedAtEpochMs: Int64, receivedAtElapsedMs: Int64, status: NetworkStatus) -> Bool {
        let event: [String: Any] = [
            "type": "network_status",
            "recorded_at_epoch_ms": recordedAtEpochMs,
            "received_at_elapsed_ms": receivedAtElapsedMs,
            "network_status": status.toJsonDictionary()
        ]
        return appendEvent(event)
    }

    func stopSession(reason: String) -> RecordingSessionInfo? {
        guard let session = activeSession else { return nil }
        let stoppedAtEpochMs = Int64(Date().timeIntervalSince1970 * 1000)
        let endEvent: [String: Any] = [
            "type": "session_end",
            "recorded_at_epoch_ms": stoppedAtEpochMs,
            "reason": reason
        ]
        try? writeJsonLine(handle: session.handle, object: endEvent)
        try? session.handle.close()
        activeSession = nil

        return RecordingSessionInfo(
            fileName: session.file.lastPathComponent,
            absolutePath: session.file.path,
            startedAtEpochMs: session.startedAtEpochMs
        )
    }

    func listSessions(limit: Int = 10) -> [SavedSessionSummary] {
        let fm = FileManager.default
        guard let files = try? fm.contentsOfDirectory(
            at: sessionsDirectory,
            includingPropertiesForKeys: [.contentModificationDateKey, .fileSizeKey, .isRegularFileKey]
        ) else {
            return []
        }

        let summaries: [SavedSessionSummary] = files.compactMap { url in
            let values = try? url.resourceValues(forKeys: [.contentModificationDateKey, .fileSizeKey, .isRegularFileKey])
            guard values?.isRegularFile == true, url.pathExtension == "jsonl" else { return nil }
            let modifiedEpochMs = Int64((values?.contentModificationDate?.timeIntervalSince1970 ?? 0) * 1000)
            let size = Int64(values?.fileSize ?? 0)
            return SavedSessionSummary(
                fileName: url.lastPathComponent,
                absolutePath: url.path,
                sizeBytes: size,
                modifiedAtEpochMs: modifiedEpochMs
            )
        }
        return Array(summaries.sorted(by: { $0.modifiedAtEpochMs > $1.modifiedAtEpochMs }).prefix(limit))
    }

    func getSessionFile(fileName: String) -> URL? {
        let candidate = sessionsDirectory.appendingPathComponent(fileName)
        var isDir: ObjCBool = false
        guard FileManager.default.fileExists(atPath: candidate.path, isDirectory: &isDir), !isDir.boolValue else {
            return nil
        }
        return candidate
    }

    private func appendEvent(_ event: [String: Any]) -> Bool {
        guard let session = activeSession else { return false }
        do {
            try writeJsonLine(handle: session.handle, object: event)
            return true
        } catch {
            return false
        }
    }

    private func writeJsonLine(handle: FileHandle, object: [String: Any]) throws {
        let data = try JSONSerialization.data(withJSONObject: object, options: [])
        handle.write(data)
        handle.write(Data([0x0A])) // newline
    }

    private struct ActiveSession {
        let file: URL
        let handle: FileHandle
        let startedAtEpochMs: Int64
    }

    enum RecorderError: Error {
        case alreadyActive
        case cannotOpenFile
    }
}

private extension ImuSample {
    func toJsonDictionary() -> [String: Any] {
        [
            "version": version,
            "sensor_id": sensorId,
            "seq": seq,
            "timestamp_ms": timestampMs,
            "ax": ax,
            "ay": ay,
            "az": az,
            "gx": gx,
            "gy": gy,
            "gz": gz
        ]
    }
}

private extension NetworkStatus {
    func toJsonDictionary() -> [String: Any] {
        [
            "version": version,
            "uptime_ms": uptimeMs,
            "active_sensor_count": activeSensorCount,
            "sensors": sensors.map { $0.toJsonDictionary() }
        ]
    }
}

private extension ActiveSensorStatus {
    func toJsonDictionary() -> [String: Any] {
        [
            "sensor_id": sensorId,
            "seq": seq,
            "age_ms": ageMs
        ]
    }
}
